### PR TITLE
Data connector KQL validations

### PR DESCRIPTION
**Pull request by mkchiliveri**
Required items, please complete
   
   Change(s):
   - Added validations for Data connector KQL validations

   Reason for Change(s):
   - Added validations for Data connector KQL validations

   Version Updated:
   - N/A

   Testing Completed:
   - Yes

   Checked that the validations are passing and have addressed any issues that are present:
   - Yes

### CODE REVIEW

File: .script/tests/KqlvalidationsTests/JsonFilesTestData/JsonFilesLoader.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    public abstract class JsonFilesLoader
+    {
+        protected const int TestFolderDepth = 6;
+
+        protected abstract List<string> GetDirectoryPaths();
+
+        public virtual List<string> GetFilesNames()
+        {
+            var directoryPaths = GetDirectoryPaths();
+            return directoryPaths.Aggregate(new List<string>(), (accumulator, directoryPath) =>
```
Comment (daxesht) [inline .script/tests/KqlvalidationsTests/JsonFilesTestData/JsonFilesLoader.cs:16]: Do we need directoryPaths null check?

File: .script/tests/KqlvalidationsTests/KqlValidationTests.cs [hunk 2]
```diff
@@ -88,18 +130,18 @@ public void Validate_HuntingQueries_SkippedTemplatesDoNotHaveValidKql(string fil
...
-        
+            var id = (string)res["id"];
+
             //Templates that are in the skipped templates should not pass the validation (if they pass, why skip?)
             if (ShouldSkipTemplateValidation(id) && res.ContainsKey("query"))
             {
-                var queryStr =  (string) res["query"];
+                var queryStr = (string)res["query"];
                 var validationRes = _queryValidator.ValidateSyntax(queryStr);
                 Assert.False(validationRes.IsValid, $"Template Id:{id} is valid but it is in the skipped validation templates. Please remove it from the templates that are skipped since it is valid.");
```
Comment (daxesht) [inline .script/tests/KqlvalidationsTests/KqlValidationTests.cs:140]: Can we consider to put message in const and use it? I can see it has been used at multiple places.

**Comment by v-atulyadav** [comment]
Hi @mkchiliveri, please have a look on comments from @daxesht, also check for branch conflicts. Thanks

**Comment by v-vdixit** [comment]
Hi @mkchiliveri, please address the comments given by @daxesht, thanks!

**Comment by v-vdixit** [comment]
Hi @mkchiliveri please provide update on this, thanks!

**Comment by v-vdixit** [comment]
Hi @mkchiliveri please provide updates on this, thanks!

**Comment by v-vdixit** [comment]
Hi @mkchiliveri please provide update on this, thanks!

**Comment by v-vdixit** [comment]
Hi @mkchiliveri please provide update on this, thanks!

**Comment by v-vdixit** [comment]
Hi @mkchiliveri please provide update on this, thanks!

**Comment by v-vdixit** [comment]
Hi @mkchiliveri we are waiting for your update on this, thanks!

**Comment by v-vdixit** [comment]
Hi @mkchiliveri we are waiting for your update on this, thanks!

**Comment by v-vdixit** [comment]
Hi @mkchiliveri we are waiting for your update on this, thanks!

**Comment by v-vdixit** [comment]
Hi @mkchiliveri we are waiting for your update on this, thanks!

**Comment by v-vdixit** [comment]
Hi @mkchiliveri please provide update on this, thanks!

File: .script/tests/KqlvalidationsTests/JsonFilesTestData/DataConnectorFilesLoader.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            var basePath = Utils.GetTestDirectory(TestFolderDepth);
+            var solutionDirectories = Path.Combine(basePath, "Solutions");
+            var dataconnectorDir = Directory.GetDirectories(solutionDirectories, "Data Connectors", SearchOption.AllDirectories);
+
+            return dataconnectorDir.ToList();
+        }
+
+        //over ride GetFilesNames method
+        public override List<string> GetFilesNames()
+        {
```
[the remark this answers is not in the document]
Comment (mkchiliveri) [reply .script/tests/KqlvalidationsTests/JsonFilesTestData/DataConnectorFilesLoader.cs:25]: Enhanced the exception handling for the method to cover all the scenarios.

File: .script/tests/KqlvalidationsTests/JsonFilesTestData/DataConnectorFilesLoader.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            var directoryPaths = GetDirectoryPaths();
+            return directoryPaths.Aggregate(new List<string>(), (accumulator, directoryPath) =>
+            {
+                var files = Directory.GetFiles(directoryPath, "*.json", SearchOption.AllDirectories).ToList();
+                List<string> validFiles = new List<string>();
+                files.ForEach(filePath =>
```
[the remark this answers is not in the document]
Comment (mkchiliveri) [reply .script/tests/KqlvalidationsTests/JsonFilesTestData/DataConnectorFilesLoader.cs:31]: Added the null check

File: .script/tests/KqlvalidationsTests/JsonFilesTestData/DataConnectorFilesLoader.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+                    var jsonString = File.ReadAllText(filePath);
+                    try
+                    {
+                        JObject dataConnectorJsonObject = JObject.Parse(jsonString);
+                        if (dataConnectorJsonObject.IsValid(dataConnectorJsonSchema))
+                        {
+                            validFiles.Add(filePath);
+                        }
+                    }
+                    catch (Exception)
```
[the remark this answers is not in the document]
Comment (mkchiliveri) [reply .script/tests/KqlvalidationsTests/JsonFilesTestData/DataConnectorFilesLoader.cs:44]: Added the exception message to console.

File: .script/tests/KqlvalidationsTests/JsonFilesTestData/DataConnectorSchema.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
+﻿using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Kqlvalidations.Tests
+{
+    public partial class DataConnectorSchema
```
[the remark this answers is not in the document]
Comment (mkchiliveri) [reply .script/tests/KqlvalidationsTests/JsonFilesTestData/DataConnectorSchema.cs:8]: removed as it is not needed.

File: .script/tests/KqlvalidationsTests/JsonFilesTestData/JsonFilesLoader.cs (hunk 1 above)
Comment (mkchiliveri) [reply .script/tests/KqlvalidationsTests/JsonFilesTestData/JsonFilesLoader.cs:16]: added null check.

File: .script/tests/KqlvalidationsTests/KqlValidationTests.cs (hunk 2 above)
Comment (mkchiliveri) [reply .script/tests/KqlvalidationsTests/KqlValidationTests.cs:140]: Updated the code with const and used it in all the methods.

**Review by daxesht — APPROVED** [review]

**Review by v-dvedak — APPROVED** [review]

**Pull request merged**